lib/active_admin/filters/active_filter.rb

```diff
@@ -75,13 +89,34 @@ def find_class?
       def find_class
         if condition_attribute.klass != resource_class && condition_attribute.klass.primary_key == name.to_s
           condition_attribute.klass
-        else
-          assoc = condition_attribute.klass.reflect_on_all_associations.
-            reject { |r| r.options[:polymorphic] }. #skip polymorphic
-            detect { |r| r.foreign_key.to_s == name.to_s }
-          assoc.class_name.constantize if assoc
+        elsif predicate_association
+          predicate_association.class_name.constantize
+        end
+      end
+
+      def filter
+        resource.filters[name.to_sym]
+      end
+
+      def related_primary_key
+        if predicate_association
+          predicate_association.active_record_primary_key
+        elsif related_class
+          related_class.primary_key
         end
       end
+
+      def predicate_association
+        @predicate_association ||= find_predicate_association
```

Comment:
@wspurgin  since find_predicate_association can return nil, 

method can look like
```
return @predicate_association if defined? @predicate_association
@predicate_association ||= find_predicate_association
```

Reply:
Gotcha. That way we don't iterate every time if their is no predicate association. I'll make that update.

Reply:
`=` is sufficient if you know it's not defined.  Alternately:
```ruby
        @predicate_association = find_predicate_association unless defined?(@predicate_association)
        @predicate_association
```